per-RPC timeouts for rpc_sync and rpc_async

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34650 per-RPC timeouts for rpc_sync and rpc_async**
timedelta
* **#34650 per-RPC timeouts for rpc_sync and rpc_async**

This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. 
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

rohan-varma committed:
[resubmit][rpc] per-RPC timeouts for rpc_sync and rpc_async

Resubmit of https://github.com/pytorch/pytorch/pull/33840, which was overly eager in the sense that it deleted a lot of code that we didn't want to get rid of yet (default timeout handling).

This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. One more issue is that torchscript doesn't support the timedelta object. This will be done in a follow up PR as it requires a fair amount of changes to the argument parsing logic.
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[resubmit][rpc] per-RPC timeouts for rpc_sync and rpc_async"


Resubmit of https://github.com/pytorch/pytorch/pull/33840, which was overly eager in the sense that it deleted a lot of code that we didn't want to get rid of yet (default timeout handling).

This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. One more issue is that torchscript doesn't support the timedelta object. This will be done in a follow up PR as it requires a fair amount of changes to the argument parsing logic. We should also discuss whether or not we want to make this change if rpc.remote() and rpc.rpc_sync() are not yet callable from JIT.
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

rohan-varma commented:
test_dist_optim is known to be flaky: https://github.com/pytorch/pytorch/issues/34141

mrshenli reviewed (commented):
Do we aim to land this for v1.5?

rohan-varma committed:
Update on "[resubmit][rpc] per-RPC timeouts for rpc_sync and rpc_async"


Resubmit of https://github.com/pytorch/pytorch/pull/33840, which was overly eager in the sense that it deleted a lot of code that we didn't want to get rid of yet (default timeout handling).

This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. One more issue is that torchscript doesn't support the timedelta object. This will be done in a follow up PR as it requires a fair amount of changes to the argument parsing logic. We should also discuss whether or not we want to make this change if rpc.remote() and rpc.rpc_sync() are not yet callable from JIT.
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

rohan-varma commented:
> Do we aim to land this for v1.5?

@mrshenli It would be good to get it in, however, if we don't have strong agreement on the API and the story for internal messages timing out and surfacing that to the user, then maybe we should not include it as part of the release. Thoughts?

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -543,6 +551,14 @@ def rpc_async(to, func, args=None, kwargs=None):
         args (tuple): the argument tuple for the ``func`` invocation.
         kwargs (dict): is a dictionary of keyword arguments for the ``func``
                        invocation.
+        timeout (Optional[datetime.timedelta]): timeout to use for this RPC. If the RPC
```
I just checked the forward for other APIs, looks like this should be 

```python
timeout (datetime.timedelta, optional)
```

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -543,6 +551,14 @@ def rpc_async(to, func, args=None, kwargs=None):
+                                                time, an exception indicating it has
+                                                timed out will be raised. If not
+                                                provided, the default value set during
+                                                initialization or with _set_rpc_timeout
```
I saw `_set_rpc_timeout` is removed from `api.py`, should we still mention it here?

rohan-varma replied:
I didn't remove it in this PR, it still exists in `rpc/init.cpp`. As you mentioned, I think that we will still need this to control timeouts for internal messages.

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -543,6 +551,14 @@ def rpc_async(to, func, args=None, kwargs=None):
+                                                time, an exception indicating it has
+                                                timed out will be raised. If not
+                                                provided, the default value set during
+                                                initialization or with _set_rpc_timeout
```
If yes, would need double quote around it.

mrshenli reviewed (commented):
Code looks good to me. What is the impact on RPC retries?

rohan-varma commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -373,15 +377,17 @@ If the future completes with an error, an exception is thrown.
       py::arg("dst"),
       py::arg("pickledPythonUDF"),
       py::arg("tensors"),
-      py::arg("rf") = nullptr);
```
This is always called with a RecordFunction (`None` if there is no profiling), so just decided to remove it. This will be going away with https://github.com/pytorch/pytorch/pull/35055 as well

rohan-varma commented:
> Code looks good to me. What is the impact on RPC retries?

There should be no impact, since this timeout is for user-facing RPCs or other RPCs that use the `::send()` API. Since retries currently work for internal messages and will retry the RPC upon failures such as a timeout, there should be no change since just the timeout will be different. @osalpekar Is this correct?

rohan-varma commented on torch/csrc/distributed/autograd/utils.h:
```diff
@@ -50,7 +50,9 @@ sendMessageWithAutograd(
-        nullptr);
+        nullptr,
+    const std::chrono::milliseconds& rpcTimeout =
+        torch::distributed::rpc::kUnsetRpcTimeout);
```
IIUC, if we use `py::none` here then we'd need to change all these APIs to accept `py::object` ? Or would we use `obj.cast<std::chrono::milliseconds>` somewhere? If we use the later, then in the case of getting `py::none`, we would still need a way to signify that no timeout was provided, correct?

osalpekar commented:
> > Code looks good to me. What is the impact on RPC retries?
> 
> There should be no impact, since this timeout is for user-facing RPCs or other RPCs that use the `::send()` API. Since retries currently work for internal messages and will retry the RPC upon failures such as a timeout, there should be no change since just the timeout will be different. @osalpekar Is this correct?

As long as we plan to tune the timeouts for user-facing RPCs only, this should be fine. I would just ensure that the retry tests in `ProcessGroupAgentTest` work as well as the `faulty_agent` tests in `dist_autograd_test.py` and `rpc_test.py`.

rohan-varma commented on torch/csrc/distributed/rpc/torchscript_functions.h:
```diff
@@ -23,7 +23,11 @@ c10::intrusive_ptr<c10::ivalue::Future> TORCH_API rpcTorchscript(
-    std::vector<c10::IValue>& stack);
+    std::vector<c10::IValue>& stack,
+    const std::chrono::milliseconds& rpcTimeout =
+        torch::distributed::rpc::kUnsetRpcTimeout); // TODO: default argument
```
Filed https://github.com/pytorch/pytorch/issues/35989

rohan-varma committed:
Update on "[resubmit][rpc] per-RPC timeouts for rpc_sync and rpc_async"


Resubmit of https://github.com/pytorch/pytorch/pull/33840, which was overly eager in the sense that it deleted a lot of code that we didn't want to get rid of yet (default timeout handling).

This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. One more issue is that torchscript doesn't support the timedelta object. This will be done in a follow up PR as it requires a fair amount of changes to the argument parsing logic. We should also discuss whether or not we want to make this change if rpc.remote() and rpc.rpc_sync() are not yet callable from JIT.
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[resubmit][rpc] per-RPC timeouts for rpc_sync and rpc_async"


Resubmit of https://github.com/pytorch/pytorch/pull/33840, which was overly eager in the sense that it deleted a lot of code that we didn't want to get rid of yet (default timeout handling).

This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. One more issue is that torchscript doesn't support the timedelta object. This will be done in a follow up PR as it requires a fair amount of changes to the argument parsing logic. We should also discuss whether or not we want to make this change if rpc.remote() and rpc.rpc_sync() are not yet callable from JIT.
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[resubmit][rpc] per-RPC timeouts for rpc_sync and rpc_async"


Resubmit of https://github.com/pytorch/pytorch/pull/33840, which was overly eager in the sense that it deleted a lot of code that we didn't want to get rid of yet (default timeout handling).

This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. One more issue is that torchscript doesn't support the timedelta object. This will be done in a follow up PR as it requires a fair amount of changes to the argument parsing logic. We should also discuss whether or not we want to make this change if rpc.remote() and rpc.rpc_sync() are not yet callable from JIT.
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

xush6528 commented on torch/distributed/rpc/api.py:
```diff
@@ -572,6 +580,14 @@ def rpc_async(to, func, args=None, kwargs=None):
         args (tuple): the argument tuple for the ``func`` invocation.
         kwargs (dict): is a dictionary of keyword arguments for the ``func``
                        invocation.
+        timeout (datetime.timedelta, optional): timeout to use for this RPC. If the RPC
```
This has to take an int type in order to work with TorchScript.

rohan-varma replied:
From an API perspective, would it then make sense to have it accept either type, and then in the python code (here) convert to timedelta if we got an int? Or should we do only `int` for the timeout in all cases? cc @mrshenli @pritamdamania87

pritamdamania87 replied:
I think we should have only one type. Although, instead of int can we use float similar to what `time.sleep()` does? This way users can specify a much more granular timeout. Internally we can convert to an appropriate int for milliseconds.

jjlilley replied:
+1 Agree float style is somewhat more intuitive for timeouts. 
I've seen different apis specify int timeouts in various units - millis, micros, seconds, etc. and can be a real source of bugs.

rohan-varma committed:
Update on " per-RPC timeouts for rpc_sync and rpc_async"


Resubmit of https://github.com/pytorch/pytorch/pull/33840, which was overly eager in the sense that it deleted a lot of code that we didn't want to get rid of yet (default timeout handling).

This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. One more issue is that torchscript doesn't support the timedelta object. This will be done in a follow up PR as it requires a fair amount of changes to the argument parsing logic. We should also discuss whether or not we want to make this change if rpc.remote() and rpc.rpc_sync() are not yet callable from JIT.
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

jjlilley commented:
Wondering: why use const& to std::chrono::milliseconds?
My understanding is that sizeof(std::chrono::milliseconds) is 8 bytes, just like a pointer.
In that case, why add any indirection?

rohan-varma commented:
> Wondering: why use const& to std::chrono::milliseconds?
> My understanding is that sizeof(std::chrono::milliseconds) is 8 bytes, just like a pointer.
> In that case, why add any indirection?

@jjlilley I wanted to try to avoid a copy. Is your suggestion that regardless we'd still use 8 bytes, and thus the reference doesn't help?

jjlilley commented:
> > Wondering: why use const& to std::chrono::milliseconds?
> > My understanding is that sizeof(std::chrono::milliseconds) is 8 bytes, just like a pointer.
> > In that case, why add any indirection?
> 
> @jjlilley I wanted to try to avoid a copy. Is your suggestion that regardless we'd still use 8 bytes, and thus the reference doesn't help?

Right, the underlying representation is effectively just an int64_t, so copying it is effectively as cheap as a pointer.

So, adding the reference makes things more expensive because of the additional deref cost.

rohan-varma committed:
Update on " per-RPC timeouts for rpc_sync and rpc_async"


Resubmit of https://github.com/pytorch/pytorch/pull/33840, which was overly eager in the sense that it deleted a lot of code that we didn't want to get rid of yet (default timeout handling).

This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. One more issue is that torchscript doesn't support the timedelta object. This will be done in a follow up PR as it requires a fair amount of changes to the argument parsing logic. We should also discuss whether or not we want to make this change if rpc.remote() and rpc.rpc_sync() are not yet callable from JIT.
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -270,8 +273,13 @@ If the future completes with an error, an exception is thrown.
                   the thread-pool used by ``ProcessGroupAgent`` (default: 4).
-              rpc_timeout (datetime.timedelta, optional): The timeout for RPC
-                  requests (default: ``timedelta(seconds=60)``).
+              rpc_timeout (datetime.timedelta, optional): The default timeout
```
Since we are now doing timeouts with `float` for torchscript reasons, should we change these to also be `float` instead of `timedelta`? If so, I can send a follow up PR doing this. We would have to change the `RpcBackendOptions` and `_set_rpc_timeout` APIs. cc @pritamdamania87 @mrshenli

mrshenli replied:
I'd vote for yes, let's make them consistent.

rohan-varma replied:
Will address in the next diff.

rohan-varma committed:
Update on " per-RPC timeouts for rpc_sync and rpc_async"


Resubmit of https://github.com/pytorch/pytorch/pull/33840, which was overly eager in the sense that it deleted a lot of code that we didn't want to get rid of yet (default timeout handling).

This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. One more issue is that torchscript doesn't support the timedelta object. This will be done in a follow up PR as it requires a fair amount of changes to the argument parsing logic. We should also discuss whether or not we want to make this change if rpc.remote() and rpc.rpc_sync() are not yet callable from JIT.
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

rohan-varma committed:
Update on " per-RPC timeouts for rpc_sync and rpc_async"


This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. 
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -58,7 +58,10 @@ PyObject* rpc_init(PyObject* /* unused */) {
               R"(URL specifying how to initialize the process group.
                 Default is ``env://``)");
 
+  // The following C++ constnats need to be cast so they can be used from
```
constnats -> constants

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -325,9 +326,14 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
+    // for all RPCs.
+    auto timeout = rpcTimeout == kUnsetRpcTimeout
+        ? getRpcTimeout()
+        : std::chrono::milliseconds((int)rpcTimeout);
```
curious, what is the reason for passing in float but then cost it to int here?

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -520,6 +520,13 @@ def rpc_sync(to, func, args=None, kwargs=None):
         args (tuple): the argument tuple for the ``func`` invocation.
         kwargs (dict): is a dictionary of keyword arguments for the ``func``
                        invocation.
+        timeout (float, optional): timeout to use for this RPC. If the RPC
```
we need to say what is the time unit for this timeout.

rohan-varma replied:
Added

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -591,6 +598,14 @@ def rpc_async(to, func, args=None, kwargs=None):
         args (tuple): the argument tuple for the ``func`` invocation.
         kwargs (dict): is a dictionary of keyword arguments for the ``func``
                        invocation.
+        timeout (float, optional): timeout to use for this RPC. If the RPC
```
ditto

rohan-varma replied:
Added

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -520,6 +520,13 @@ def rpc_sync(to, func, args=None, kwargs=None):
         args (tuple): the argument tuple for the ``func`` invocation.
         kwargs (dict): is a dictionary of keyword arguments for the ``func``
                        invocation.
+        timeout (float, optional): timeout to use for this RPC. If the RPC
```
shall we also mention that 0 timeout means infinity?

rohan-varma replied:
Good point, added

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1694,19 +1699,62 @@ def test_rpc_timeouts(self):
         fut1.wait()
 
-        # future should run to completion if the timeout is zero.
+        # Zero timeout means infinity, so future should ru to completion.
```
ru -> run ?

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1694,19 +1699,62 @@ def test_rpc_timeouts(self):
+    def test_rpc_timeouts(self):
+        # TODO: enable timeouts for rpc.remote/RRef (https://github.com/pytorch/pytorch/issues/33803)
+        dst_rank = (self.rank + 1) % self.world_size
+        dst_worker = "worker{}".format(dst_rank)
```
rpc_test.py now imports a `worker_name` helper function that does this.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1694,19 +1699,62 @@ def test_rpc_timeouts(self):
+        sec_to_ms = 1000
+        rpc.rpc_async(dst_worker, my_sleep_func, args=(1,), timeout=5 * sec_to_ms).wait()
+        rpc.rpc_sync(dst_worker, my_sleep_func, args=(1,), timeout=5 * sec_to_ms)
+        # Passing in a zero timeout should ensure that the RPC wont time out.
```
wont -> won't

mrshenli reviewed (approved):
LGTM. Most of my comments are doc-related or cosmetic, expect one question. What is the reason for using float for timeout but cast to int internally?

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -292,7 +292,8 @@ void ProcessGroupAgent::shutdownImpl() {
     const WorkerInfo& to,
-    Message&& message) {
+    Message&& message,
+    const float rpcTimeout) {
```
not required: I've found it can be helpful in some cases to add the unit to the variable name in apis.
e.g. rpcTimeoutSeconds or rpcTimeoutMilliseconds

btw, if it is seconds (which is common for float), you might want to multiply by 1000 below before casting

rohan-varma replied:
I originally meant for this to be in ms, although, not sure what was intended when the `float` type was suggested (cc @pritamdamania87 )

jjlilley replied:
fwiw, in many/most cases where I've seen timeouts represented as a `float`, the unit is seconds.

Whatever we got with, adding units to the var name might help self-document and eliminate confusion. I've seen a ton of seconds/msec/usec mixups...

rohan-varma replied:
@jjlilley Updated to be seconds and changed the param to `rpcTimeoutSeconds`

rohan-varma commented:
> LGTM. Most of my comments are doc-related or cosmetic, expect one question. What is the reason for using float for timeout but cast to int internally?

I think the cast is needed due to the required constructor arg needing to be an int, i.e. `const std::chrono::milliseconds(float)` is not valid.

mrshenli commented:
> I think the cast is needed due to the required constructor arg needing to be an int, i.e. const std::chrono::milliseconds(float) is not valid.

Is there any reason that we cannot take a int from Python API directly?

pritamdamania87 commented:
> Is there any reason that we cannot take a int from Python API directly?

My suggestion to use float was mostly inspired from something like `time.sleep()` which uses a float in seconds. With a float users can specify time in much lower granularity than what we can support via int. For example, if the timeout is int and is in milliseconds, users can't go below 1ms (with float we can always change the internal implementation to use std::chrono::microseconds/nanoseconds if needed). Although, it's unlikely someone would need to set a timeout lower than 1ms in a RPC framework. I'm fine with using an int if we document the units clearly and it is milliseconds.

rohan-varma committed:
Update on " per-RPC timeouts for rpc_sync and rpc_async"


This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. 
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

rohan-varma commented:
Decided to go with the original idea of float + seconds unit. Updated documentation and variable names to indicate so

rohan-varma committed:
Update on " per-RPC timeouts for rpc_sync and rpc_async"


This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. 
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

rohan-varma committed:
Update on " per-RPC timeouts for rpc_sync and rpc_async"


This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. 
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

rohan-varma committed:
Update on " per-RPC timeouts for rpc_sync and rpc_async"


This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. 
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

jjlilley reviewed (approved):
Thanks for the changes - fwiw, LGTM!

Aside: we might want to change the units of the process_group_agent-wide timeout float/seconds for consistency, but best a separate self-contained change.

rohan-varma committed:
Update on " per-RPC timeouts for rpc_sync and rpc_async"


This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. 
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

rohan-varma committed:
Update on " per-RPC timeouts for rpc_sync and rpc_async"


This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. 
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

rohan-varma committed:
Update on " per-RPC timeouts for rpc_sync and rpc_async"


This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. 
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

rohan-varma committed:
Update on " per-RPC timeouts for rpc_sync and rpc_async"

timedelta
* **#34650 per-RPC timeouts for rpc_sync and rpc_async**

This PR adds an optional argument into `rpc_sync` and `rpc_async` as well as `RpcAgent::send()` that allows the user to specify a timeout for an RPC to override the default set timeout. If the user does not specify this argument, then the currently set default RPC timeout given in the RPC constructor or by `rpc.set_rpc_timeout()` is used. Otherwise, we use the passed in timeout.

This diff does not address:
1) timeout support when called rpc.rpc_async is called as a JIT operator. For this to work, we would need to change the logic in `register_distributed_ops` to pass in this timeout to `rpcTorchscript`. 
2) Per-RPC timeouts for internal messages or `rpc.remote()`. A follow-up diff will address the latter with the approach of raising the timeout error at the earliest next possible time to the user, such as when the next time the RRef is forked or `to_here` is called

Added unit tests to confirm the current behavior

Differential Revision: [D20376953](https://our.internmc.facebook.com/intern/diff/D20376953/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D20376953/)!

[ghstack-poisoned]

rohan-varma commented:
> Thanks for the changes - fwiw, LGTM!
> 
> Aside: we might want to change the units of the process_group_agent-wide timeout float/seconds for consistency, but best a separate self-contained change.

Addressing this change in https://github.com/pytorch/pytorch/pull/37027
